src/main/scala/z3/scala/dsl/ValHandler.scala

```diff
@@ -20,9 +20,9 @@ abstract class ValHandler[A : Default] {
   /** Z3 code to construct a sort representing the Scala A type. */
   def mkSort(z3 : Z3Context) : Z3Sort
 
-  /** Constructs a Val[A], i.e. the representation of a variable of type A. */
-  def construct : Val[A] = new Val[A] {
-    def build(z3 : Z3Context) = z3.mkFreshConst("valCst", mkSort(z3))
+  /** Constructs a ValTree[A], i.e. the representation of a variable of type A. */
+  def construct[T >: BottomSort <: TopSort]: Tree[T] = new ValTree[T] {
```

Comment:
Yes, I think that would unfortunately be the case. I like your unification of the glue code idea better too :) In the end, we may however need to rely on explicitly writting the implicit parameters since we do in fact need two distinct type parameters for this to work.